File: scripts/playlist_to_tracks.py
# scripts/playlist_to_tracks.py
import os, json, requests
from pathlib import Path
from xml.etree import ElementTree as ET

PLAYLIST_ID = os.environ.get("PLAYLIST_ID") or input("Enter playlist ID: ").strip()
if not PLAYLIST_ID:
    raise SystemExit("No PLAYLIST_ID provided")

RSS_URL = f"https://www.youtube.com/feeds/videos.xml?playlist_id={PLAYLIST_ID}"
headers = {"User-Agent": "Mozilla/5.0"}

resp = requests.get(RSS_URL, headers=headers, timeout=20)
resp.raise_for_status()

# Namespaces: Atom (default), YouTube, Media RSS
ns = {
    "atom": "http://www.w3.org/2005/Atom",
    "yt": "http://www.youtube.com/xml/schemas/2015",
    "media": "http://search.yahoo.com/mrss/",
}

root = ET.fromstring(resp.text)
entries = root.findall("atom:entry", ns)  # ← correct
if not entries:
    print("No <entry> elements found. Is the playlist public and ID correct?")
    # helpful debug:
    feed_title = root.findtext("atom:title", default="(no feed title)", namespaces=ns)
    print("Feed title:", feed_title)
    # bail without writing an empty file
    raise SystemExit(1)

tracks = []
for idx, entry in enumerate(entries):
    title = entry.findtext("atom:title", default="", namespaces=ns).strip()
    vid_el = entry.find("yt:videoId", ns)
    video_id = (vid_el.text if vid_el is not None else "").strip()
    # media:description is inside media:group
    desc_el = entry.find("media:group/media:description", ns)
    description = (desc_el.text or "").strip() if desc_el is not None else ""
    # optional: thumbnail + publishedAt
    thumb_el = entry.find("media:group/media:thumbnail", ns)
    thumbnail = thumb_el.attrib.get("url") if thumb_el is not None else None
    published_at = entry.findtext("atom:published", default="", namespaces=ns)

    tracks.append({
        "title": title,
        "id": video_id,
        "description": description,
        "tags": [],
        "slug": f"/lore#log-{str(idx).zfill(3)}",
        "cover": thumbnail,
        "publishedAt": published_at,
    })

# Write to project-root/src/data/tracks.json
project_root = Path(__file__).resolve().parents[1]
out_path = project_root / "src" / "data" / "tracks.json"
out_path.parent.mkdir(parents=True, exist_ok=True)
with open(out_path, "w", encoding="utf-8") as f:
    json.dump(tracks, f, indent=2)

print(f"Wrote {len(tracks)} items → {out_path}")

File: scripts/playlist_to_tracks.js
// Usage: node scripts/playlist_to_tracks.js <PLAYLIST_ID>
// Converts YouTube playlist RSS into src/data/tracks.json (no API key).
import fs from 'node:fs';
import path from 'node:path';
import https from 'node:https';

const playlistId = process.argv[2];
if (!playlistId) {
  console.error('Provide a PLAYLIST_ID. Example: node scripts/playlist_to_tracks.js PLabcd1234');
  process.exit(1);
}
const rssUrl = `https://www.youtube.com/feeds/videos.xml?playlist_id=${playlistId}`;

function fetch(url) {
  return new Promise((resolve, reject) => {
    https.get(url, res => {
      let data = '';
      res.on('data', d => (data += d));
      res.on('end', () => resolve({ status: res.statusCode, text: () => data }));
    }).on('error', reject);
  });
}
function xmlTag(str, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\s\S]*?)</${tag}>`, 'i');
  const m = str.match(re);
  return m ? m[1] : '';
}
function xmlTags(str, tag) {
  const re = new RegExp(`<${tag}[^>]*>[\s\S]*?</${tag}>`, 'ig');
  return [...str.matchAll(re)].map(m => m[0]);
}
function getAttr(tagStr, attr, ns=false) {
  const re = new RegExp(`${ns?attr.replace(':','\\:'):attr}="([^"]+)"`, 'i');
  const m = tagStr.match(re);
  return m ? m[1] : '';
}
function strip(s) {
  return s.replace(/<!\[CDATA\[|\]\]>/g,'').trim();
}

const OUT = path.resolve('src/data/tracks.json');

(async () => {
  const res = await fetch(rssUrl);
  if (res.status !== 200) {
    console.error('Failed to fetch RSS:', res.status);
    process.exit(1);
  }
  const rss = await res.text();

  const entries = xmlTags(rss, 'entry').map(e => {
    const title = strip(xmlTag(e, 'title'));
    const videoId = getAttr(e, 'yt:videoId', true) || strip(xmlTag(e, 'yt:videoId'));
    const desc = strip(xmlTag(e, 'media:description')) || '';
    return { title, id: videoId, description: desc };
  });

  const items = entries.map((it, idx) => ({
    title: it.title,
    id: it.id,
    description: it.description,
    tags: [],
    slug: `/lore#log-${String(idx).padStart(3,'0')}`
  }));

  fs.mkdirSync(path.dirname(OUT), { recursive: true });
  fs.writeFileSync(OUT, JSON.stringify(items, null, 2));
  console.log(`Wrote ${items.length} items → ${OUT}`);
})();
